swift/ql/lib/codeql/swift/frameworks/StandardLibrary/WebView.qll

```diff
@@ -27,3 +32,98 @@ private class WKScriptMessageBodyInheritsTaint extends TaintInheritingContent,
     )
   }
 }
+
+/**
+ * A model for `JSContext` sources. `JSContext` acts as a bridge between JavaScript and
+ * native code, so any object obtained from it has the potential of being tainted by a malicious
+ * website visited in the WebView.
+ */
+private class JsContextSources extends SourceModelCsv {
+  override predicate row(string row) {
+    row =
+      [
+        ";JSContext;true;globalObject;;;;remote",
+        ";JSContext;true;objectAtIndexedSubscript(_:);;;ReturnValue;remote",
+        ";JSContext;true;objectForKeyedSubscript(_:);;;ReturnValue;remote"
```

Comment:
Should these things also be taint sources?
 - the value of `JSContext.exception`
 - the second input to `JSContext.exceptionHandler`?
 - the return value from both variants of `evaluateScript`?

Reply:
> the value of `JSContext.exception`

[`JSContext.exception`](https://developer.apple.com/documentation/javascriptcore/jscontext/1451499-exception) is an exception that gets thrown on the JavaScript side once the native code execution has finished. So I don't _think_ JavaScript can use that field to execute native code.

> the second input to `JSContext.exceptionHandler`?

The `exceptionHandler` is a block that executes native code indeed, but I think it can't be set from the JavaScript side. The `JSValue` that is the second argument is the exception that got thrown in a native code callback, which again doesn't seem to be JS-provided nor should affect the native code — it's targeted to JavaScript so that the exception can be handled there (if I understood [all this](https://developer.apple.com/documentation/javascriptcore/jscontext/1451731-exceptionhandler) correctly).

> the return value from both variants of `evaluateScript`?

That indeed comes for JavaScript. The thing is that the JavaScript being executed in this case comes from the app itself, not the website visited in the WebView. So unless the existing JavaScript could somehow tamper the code passed to `evaluateScript`, its return value wouldn't be under the malicious site's control. Of course controlling the script of the `evaluateScript` call itself would be a problem, but that's more akin to a XSS vulnerability.

Reply:
Thanks for clearing these things up.  It sounds like you've considered these cases in much more detail than I have!